src/compiler/checker.ts

```diff
@@ -18117,7 +18117,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     removeFromEach(typeSet, TypeFlags.Null);
                     result = getUnionType([getIntersectionType(typeSet, flags), nullType], UnionReduction.Literal, aliasSymbol, aliasTypeArguments);
                 }
-                else if (typeSet.length >= 4) {
+                else if (typeSet.length >= 3 && types.length > 2) {
                     // When we have four or more constituents, some of which are unions, we employ a "divide and conquer" strategy

```

Comment:
Should this comment be updated to accurately describe the new conditions (typeSet.length >= 3 and types.length > 2)?